Update EIP-7873: EOF - flag to include caller in Creator Contract (#9474)
ethereum/EIPs: EIPS/eip-7873.md

```solidity
{

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Author: good point, let me fix that and see if no hairy issues arise

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Author: Done 38d4a9d5, pls double check

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Reviewer: @shemnon's suggestion above implies that `if lt(size, 65) { revert(0, 0) }` should be changed to `if lt(size, 32) { revert(0, 0) }`, so that all of `[tx_initcode_hash][salt][init_data]` become optional arguments with default zero/empty.

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Reviewer: That's not sufficient though. If `32 < calldatasize < 65` you'd want to bump `size` up to 65 so that the caller is correctly appended.

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Author: I didn't take that suggestion went that far, that is, the caller would still need to explicitly send the 33 zero bytes (at a discount). That is, the advantage was more of a elegance standpoint. Re-reading now I think you're right and requiring only 32 bytes was the point.

This raises a question whether we want to rely on calldata behavior like that? It also kinda makes the whole thing more complex - as you've already kinda noticed. :thinking:

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Reviewer 2: Good point on the discouraging zero salt.  

Perhaps we should consider asking tooling that sends these deployments to use some function on the the sender's nonce for a salt unless explicitly specified?

>     /// Takes [unsafe_flag][tx_initcode_hash][salt][init_data] as input,

Author: in the other thread we landed at a conclusion that we still want the contract to require a 65-byte input to proceed. So we're keeping the proposed ordering but just for the sake of elegance of starting the input with a non-zero argument of `tx_initcode_hash`

    /// creates contract and returns the address or failure otherwise.

    /// tx_initcode_hash and salt are 32 bytes wide, unsafe_flag is 1 byte wide, init_data can be any width.
    /// If unsafe_flag is 0x00, `CALLER` is included in the salt and impacts the target address of the created account.
    /// If unsafe_flag is non-zero, `CALLER` is excluded. Deployment by any sender will result in same target address. 
    /// Deployments using non-zero unsafe_flag are susceptible to front-running.

    // init_data.length can be 0, but the first 65 bytes are mandatory
    let size := calldatasize()

>     if lt(size, 65) { revert(0, 0) }
> 
>     // copy tx_initcode_hash, salt, unsafe_flag and init_data to memory to hash.
>     // note that the unsafe_flag is included to prevent pre-image attacks on the two paths to calculate salt below.
>     calldatacopy(0, 0, size)

Reviewer: FWIW, I believe this implements @shemnon's suggestion.

```suggestion
    // revert if tx_initcode_hash not provided.
    if lt(size, 32) {
        revert(0, 0)
    }

    // copy parameters to memory to hash.
    // note that the unsafe_flag is included to prevent pre-image attacks on the two paths to calculate salt below.
    calldatacopy(0, 0, size)

    // increase size if salt or unsafe_flag not provided, defaulting to zero.
    if lt(size, 65) {
        size := 65
    }
```

Author: I had it drafted similar, but did the "zero-padding-defaulting" on the `calldatacopy` step, i.e. I expand `size` one step earlier. More expensive but maybe safer.

Anyway, by "moot" you mean to keep it as is (no zero padding by default, require 65 bytes)? I think simpler is better.

Reviewer: Yes, let's just require 65 bytes. This comment should be resolved.

Reviewer 3: agree this seems too complex without much benefit, I would require all arguments

    // mask out the 31 bytes that follow the flag in input data

>     let unsafe_flag := calldataload(64) & (0xff << 248)

Reviewer: ```suggestion
    let unsafe_flag := byte(0, calldataload(64))
```

    if iszero(unsafe_flag) {
        // Include sender into salt just after init_data
        mstore(msize(), caller())

>         let final_salt := keccak256(0, size + 32)

Reviewer: I think there is a bug here: `msize()` is rounded up to multiple of 32, but `init_data` may not be a multiple, so this code results in some padding between `init_data` and sender, and since the hash uses `size` it does not actually include the full address in the final salt.

>     if iszero(unsafe_flag) {
>         // Include sender into salt just after init_data
>         mstore(msize(), caller())
>         let final_salt := keccak256(0, size + 32)
>     } else {
>         // Exclude sender from salt - susceptible to front-running
>         let final_salt := keccak256(0, size)
>     }

Reviewer: Just an idea, this seems equivalent:

```suggestion
    if iszero(unsafe_flag) {
        // Include sender into salt just after init_data
        mstore(msize(), caller())
    }

    let final_salt := keccak256(0, msize())
```

    let tx_initcode_hash := calldataload(0)
    let init_data_size := sub(size, 65)

    // reuse init_data which has been already copied to memory above

>     let ret := txcreate(tx_initcode_hash, callvalue(), final_salt, 65, init_data_size)

Author: yeah, definitely pseudo-code. I knew this was maybe a stretch on the syntax, but don't want also to overthink it.

```solidity
{
    /// Takes [tx_initcode_hash][salt][unsafe_flag][init_data] as input,
    /// creates contract and returns the address or failure otherwise.

>     /// tx_initcode_hash and salt are 32 bytes wide, unsafe_flag is 1 byte wide, init_data can be any width.

Author: New topic: shouldn't we ABI encode the arguments and make the API more future-proof? The reason to do this would be to achieve an open standard for inputs to an EOF factory, so that any ERC-driven factories can adhere to and extend.

This would have the benefit for the tooling, which would (maybe) not need to have an entirely separate path to encode arguments for the Creator Contract and the other factories. E.g. it could abstract the choice of factory and some "extra" params out and leave it up to the user.

Now, coming up with such a standard future-proof API is a huge task, so I'm playing devil's advocate here - if we can't realistically offer that, it might be less likely Creator Contract will be used long-term.

---

A minimalist hack **just to illustrate the idea** would be to have:

```
[tx_initcode_hash][salt][unsafe_flag][abi_encoded_extra:bytes][init_data]
```

where `abi_encoded_extra:bytes` would be assumed to hold ABI encoded `bytes` holding inside opaque bytes with extra arguments to any future extensions. `abi_encoded_extra` would be required, but otherwise ignored by the Creator Contract (except decoding the length of bytes to know how much input to require and ignore). `abi_encoded_extra` would be exempt from salt hashing. Vanilla usage would be to send 32 zero bytes meaning an empty `bytes`.

Reviewer 3: I also thought about ABI, and convenience for tooling might be a good argument in favor (also, would it be easier to call from Solidity?). But devising some general API for factories is going too far, definitely out of scope.

Author: > I also thought about ABI, and convenience for tooling might be a good argument in favor (also, would it be easier to call from Solidity?). But devising some general API for factories is going too far, definitely out of scope.

What about the "flipped" argument that not having a future-proof ABI/API contradicts the notion that Creator Contract will be used in the long term (which we're basing off of to introduce `unsafe_flag` in the first place here)?

Reviewer: I really don't think that non-standard encoding will be an issue for tooling. A shared abstraction covering other kinds of factories can be built even if different factories require different encoding.

    /// If unsafe_flag is 0x00, `CALLER` is included in the salt and impacts the target address of the created account.
    /// If unsafe_flag is non-zero, `CALLER` is excluded. Deployment by any sender will result in same target address. 
    /// Deployments using non-zero unsafe_flag are susceptible to front-running.

    // init_data.length can be 0, but the first 65 bytes are mandatory
    let size := calldatasize()
    if lt(size, 65) { revert(0, 0) }

    // copy tx_initcode_hash, salt, unsafe_flag and init_data to memory to hash.
    // note that the unsafe_flag is included to prevent hash collisions between the two branches:
    //   - unsafe_flag == 0, caller's address is appended by the code

>     //   - unsafe_flag == 1, caller's address is sent via init_data

Reviewer: This is a little weird. Nothing forces the caller to include their address in `init_data`. In most cases that use non-zero unsafe flag the deployment should be totally independent of the caller. Additionally, the flag can be any non-zero value.

Reviewer 3: I think this comment means that without hashing `unsafe_flag` the deployment with `unsafe_flag == 0` collides with deployment with `unsafe_flag == 1` and caller's address in `init_data`

Reviewer: Ok, that makes more sense.

    calldatacopy(0, 0, size)

    // mask out the 31 bytes that follow the flag in input data

>     let unsafe_flag := byte(0, calldataload(64))

Author: I had another thought on this - maybe it's better to require the unsafe_flag be just one lowest bit of that byte, so that the other bits can be possibly repurposed for other flags? The contract is not upgradable (nor planned to be re-deployed by state change), but... maybe this tiny amount forward compatibility would be useful?

This of course opens another can of worms: should we revert on higher bits set?

Reviewer: The forward compatible option would be to force `unsafe_flag` to be `0x00` or `0x01` and revert on other values.

I also find that tempting. Using RJUMPV it would be very simple to express in the code. So I'd be in favor of this change.

Author: > The forward compatible option would be to force `unsafe_flag` to be `0x00` or `0x01` and revert on other values.
> 
> I also find that tempting. Using RJUMPV it would be very simple to express in the code. So I'd be in favor of this change.

Ah, right, if we revert on unsafe_flag > 0x01 then no need to worry about hashing. I find this an extra advantage of such restricted approach - otherwise unsafe_flag's higher 7 bits become de-facto salt, which is a bit weird.

@gumb0 can you elaborate on your reservations towards this (you've :-1:ed )?

Reviewer 3: To me it feels like overthinking and unnecessary logic complication for speculative benefit. Current code with branching on `iszero(unsafe_flag)` looks fine, for hashing I would do Pawel's suggestion to hash 0 address instead of caller for unsafe mode and don't hash the flag.

Author: Resolved, see other thread and let's continue there

    if iszero(unsafe_flag) {
        // Include sender into salt just after init_data
        mstore(size, caller())

>         let final_salt := keccak256(0, size + 32)

Reviewer 4: This is insecure. Even if you include the address of the caller this can be front-runned by:
- setting unsafe_flag to 1,
- appending the caller's address to the input.

I think the better approach is to have a fixed placeholder of 32-bytes for the caller's address before the input data. If `unsafe_flag` is set leave the placeholder all zeros, otherwise write caller's address there.

Reviewer: Alternatively, always hash over `size + 32` but only write the sender in the last word if `unsafe_flag` is nonzero.

Author: `unsafe_flag` is included in the hashing, so that front-running won't work (will land in different target address), and there's a comment laying out this exact reason (also raised on discord and explained).

> // note that the unsafe_flag is included to prevent hash collisions between the two branches

Am I maybe missing sth?

Reviewer: No that sounds right... If a tx is "frontrun" as @chfast suggests, it should result in a different deployed address.

Author: I pushed the proposed alternative to 0b55a4f5. Note that this is slightly more complex actually, as long as we keep the earlier suggestion to place `unsafe_flag` as third argument. We can debate whether or not to keep it there.

Overall it has a nice property of cutting out the address malleability via various non-zero values of `unsafe_flag` byte...

Author: @gumb0 @chfast LMK if you're happy with the latest revision and :heavy_check_mark: if yes, so we can merge this if there's also consensus on the overall direction.

Reviewer 4: I have more ideas in https://github.com/ipsilon/eof/issues/172#issuecomment-2732366390, but maybe better to make another PR to present my new ideas :)

So I think this is fine to merge.

    /// If unsafe_flag is 0x00, `CALLER` is included in the salt and impacts the target address of the created account.
    /// If unsafe_flag is non-zero, `CALLER` is excluded, with 32 zero-bytes used instead. Deployment by any sender will result in same target address. 
    /// Deployments using non-zero unsafe_flag are susceptible to front-running.

    // init_data.length can be 0, but the first 65 bytes are mandatory
    let size := calldatasize()
    if lt(size, 65) { revert(0, 0) }

    // copy tx_initcode_hash and salt to memory to hash.
    // note that the unsafe_flag doesn't need to be included. If set to non-zero, 32 zero-bytes
    // are used instead of the CALLER address bytes, which prevents hash collisions between
    // deployments using unsafe_flag and those not using it.
    calldatacopy(0, 0, 64)

    // mask out the 31 bytes that follow the flag in input data
    let unsafe_flag := byte(0, calldataload(64))

    if iszero(unsafe_flag) {
        // store caller in memory to hash, just after salt
        mstore(64, caller())
    } else {
        // exclude sender from salt - susceptible to front-running

>         mstore(64, 0)

Author: Just explicitness (and me being forgetful). I'll switch the mstore to just a suitable comment
